src/lib/markbind/src/lib/markdown-it/index.js

```diff
@@ -42,6 +30,72 @@ markdownIt.renderer.rules.table_close = (tokens, idx) => {
   return '</table></div>';
 };
 
+// syntax highlight code fences and add line numbers
+markdownIt.renderer.rules.fence = (tokens, idx, options, env, slf) => {
+  const token = tokens[idx];
+  const lang = token.info || '';
+  let str = token.content;
+  let highlighted = false;
+  if (lang && hljs.getLanguage(lang)) {
+    try {
+      str = hljs.highlight(lang, str).value;
+      highlighted = true;
+    } catch (__) {}
+  }
+  if (!highlighted) {
+    str = markdownIt.utils.escapeHtml(str);
+  }
+  
+  const linesToHighlight = token.attrGet('highlight-lines');
+  let lineNumbersAndRanges = [];
+  if (linesToHighlight) {
+    // example input format: "1,4-7,8,11-55"
+    //               output: [[1],[4,7],[8],[11,55]]
+    // the output is an array contaning either single line numbers [lineNum] or ranges [start, end]
+    // ',' delimits either single line numbers (eg: 1) or ranges (eg: 4-7)
+    arrLinesToHighlight = linesToHighlight.split(',');
+    // if it's the single number, it will just be parsed as an int, (eg: ['1'] --> [1] )
+    // if it's a range, it will be parsed as as an array of two ints (eg: ['4-7'] --> [4,6])
+    lineNumbersAndRanges = arrLinesToHighlight.map(elem => elem.split('-').map(lineNumber => parseInt(lineNumber, 10)));
+  }
+
+  const lines = str.split('\n');
+  lines.pop(); // last line is always a single '\n' newline, so we remove it
+  // wrap all lines with <span> so we can number them
+  str = lines.map((line, index) => {
+    // if a line is empty we put a 0 width non breaking space
+    const content = line || '&#x200B;';
+    const currentLineNumber = index + 1;
+    // check if there is at least one range or line number that matches the current line number
+    const inRange = lineNumbersAndRanges.some(([start, end]) => {
```

Comment:
I referenced the algorithm off [markdown-it-highlight-lines](https://github.com/egoist/markdown-it-highlight-lines/blob/master/src/index.js)

Reply:
That's a good point. You could bring it down to O(lines) with a more efficient algorithm here, but I don't think it would make much of a performance difference unless the user has some really massive files.

I'd towards a simpler implementation that's easier to reason about for the purpose of maintainability. If we do find this a bottleneck at some point in time, we can optimize it then 🙂

Reply:
Very well then, is there a way we can let future devs know that this is a possible area of optimization? A //todo perhaps?

Reply:
It might be worth leaving a comment explaining why we chose to go with this approach for now (simplicity). I wouldn't necessarily label it a TODO because there's no definite or expected need for us to do anything to change this.